### DISCUSSION TITLE
Unexpected fastapi + uvicorn concurrency

### DISCUSSION THREAD

**Discussion by floatingpurr** [fastapi/fastapi#4358] [Questions] [answered]
In [this example](https://stackoverflow.com/questions/70446584/how-does-fastapi-uvicorn-parallelize-requests) coming from SO, the OP defines an endpoint as simple as this one:

```python
@app.get('/loadtest')
def root():
    time.sleep(1)
    return {'message': 'hello'}
```

the app is run with `uvicorn` without any workers, but when she/he runs:

`ab -c 100 -n 1000 localhost/loadtest`

it looks like requests are processed in parallel. It looks strange because:
- `await` is not used
- Any concurrency method is defined

I'm wondering why this happens.

**Comment by nymous** [comment, marked as answer]
For endpoints defined with `def` (not `async def`), FastAPI will run them in a threadpool, exactly as to avoid blocking the server and allow multiple requests to be served in parallel.
See https://fastapi.tiangolo.com/async/#path-operation-functions for more technical details.

Note that if you define an endpoint with `async def` but run synchronous IO in it, then it *will* block the server, as FastAPI will execute it in the event loop instead of an external threadpool.

**Reply by floatingpurr** [reply]
Thanx for your reply. Regarding the threadpool, is there a way to estimate how many concurrent requests can be handled?

**Reply by raphaelauv** [reply]
@floatingpurr -> https://github.com/tiangolo/fastapi/issues/4221#issuecomment-982260467
